---
title: "3-4. trows"
date: 2023-06-04
---
## Ⅰ. throws
如果一个方法(中的语句执行时)可能生成某种异常，但是并不能确定如何处理这种异常，则此方法应显示地声明抛出异常，表明该方法将不对这些异常进行处理，而由该方法的调用者负责处理。

在方法声明中用throws语句可以声明抛出异常的列表，throws后面的异常类型可以是方法中产生的异常类型，也可以是它的父类 。

#### 1. throws介绍
throws只是用来抛出异常的，将异常抛到代码的调用处，并不会处理异常，处理异常还是要在代码调用处使用try-catch结构。

"throws + 异常类型"写在方法的声明处。指明此方法执行时，可能会抛出的异常类型。一旦当方法体执行时，出现异常，仍会在异常代码处生成一个异常类的对象，此对象满足throws后异常类型时，就会被抛出。异常代码后续的代码，就不再执行！

#### 2. 异常的抛出机制
如果一个方法内抛出异常，该异常对象会被抛给调用者方法中处理。如果异常没有在调用者方法中处理，它继续被抛给这个调用方法的上层方法。这个过程将一直继续下去，直到异常被处理。

如果一个异常回到main()方法，并且main()也不处理，则程序运行终止。

![3-4-1](/img/java/javase/3-4-1.jpg)

#### 3. throws使用示例
```java
public class ExceptionTest2 {
  public static void main(String[] args){ //这里再抛就到虚拟机了，不能再抛了
    try{ //这里开始处理异常
      method2();
    }catch(IOException e){
      e.printStackTrace();
    }
  }
  
  //因为IOException包含了FileNotFountException,所以可以合并成一个异常
  //如果这两个异常的处理方式不一样，比如返回的错误码不一样，则不要合并。
  public static void method2() throws IOException{
    method1(); //异常又被抛出去了，所以这里不用处理异常
  }
  
  //这里将异常抛出到调用处，因此不用进行处理。
  public static void method1() throws FileNotFoundException,IOException{
    File file = new File("hello1.txt");
    FileInputStream fis = new FileInputStream(file);
    
    int data = fis.read();
    while(data != -1){
      System.out.print((char)data);
      data = fis.read();
    }
    fis.close();
    System.out.println("hahaha!");
  }
}
```
#### 4. try-catch-finally与throws的对比
try-catch-finally:真正的将异常给处理掉了。

throws的方式只是将异常抛给了方法的调用者。并没有真正将异常处理掉。

#### 5. 方法重写时的异常规则
子类重写的方法抛出的异常类型不大于父类被重写的方法抛出的异常类型。

因为我们调用父类方法时要进行异常处理，而此时若使用多态，实际调用的是子类的方法，则必须能处理掉子类的异常，不然会报错。

#### 6. 开发中try-catch-finally和throws的选择
- 如果父类中被重写的方法没有throws方式处理异常，则子类重写的方法也不能使用throws，意味着如果子类重写的方法中有异常，必须使用try-catch-finally方式处理。
- 执行的方法a中，先后又调用了另外的几个方法，这几个方法是递进关系执行的。我们建议这几个方法使用throws的方式进行处理。而执行的方法a可以考虑使用try-catch-finally方式进行处理。

#### 7. 运行时异常的排查
如果代码编译时没有异常，运行时出现了异常，比如空指针；那么应该对代码进行排查，优化，而不是进行try..catch处理。


## Ⅱ. 手动抛出异常
#### 1. 关于异常的产生
- 系统自动生成的异常对象
- 手动生成一个异常对象，并抛出（throw）

#### 2. 手动抛出异常的意义
异常是抛给调用者看的，调用者分为其他方法调用，或前端调用。有些异常是逻辑上的，需要我们自己手动抛出。抛出异常和返回一个特殊值，都能解决问题；可自行选择，抛出异常可以省掉很多`if...else...` 但是返回值更精准。

如果代码存在逻辑上的问题，我们就需要抛出异常，告诉代码的调用者。

#### 3. 创建异常的语法
由于是逻辑上的异常，需要我们自己new异常类，我们选择的异常类型有：
- Exception：编译异常（编译时会报错）
- RuntimeException：运行时异常（选他，不会报错）-->选这个
```java
throw new Exception("异常信息");
throw new RuntimeException("异常信息");
```

#### 4. 示例
自定义的异常也是可以throws到调用处的，比如抛出的是个编译时异常，就可以抛出到调用处进行处理。自定义异常中定义的异常信息，可以在调用处，使用异常类的getMessage方法获取。
```java
class Student{  
  private int id;
  //这里可以将异常信息再次throws到调用处
  public void regist(int id) throws Exception {
    if(id > 0){
      this.id = id;
    }else{
       //手动抛出异常对象,运行时异常不会报编译时错误
       throw new RuntimeException("您输入的数据非法！");
       //Exception会报编译时错误，可以将该异常throws到调用处
      throw new Exception("您输入的数据非法");
    }
  }
}
 public class StudentTest {
  public static void main(String[] args) {
    try { //处理throws来的的异常
      Student s = new Student();
      s.regist(-1001);
      System.out.println(s);
    } catch (Exception e) {
      System.out.println(e.getMessage());
    }
  }
} 
```


## Ⅲ. 自定义异常类

#### 1. 自定义异常类的创建说明
- 一般地，用户自定义异常类都是RuntimeException的子类；
- 自定义异常类通常需要编写几个重载的构造器。
- 自定义异常需要提供serialVersionUID
    - 类的标识，随便定义，要是long类型，够长
- 自定义的异常通过throw抛出。
- 自定义异常最重要的是异常类的名字，当异常出现时，可以根据名字判断异常类型。

#### 2. 自定义异常类示例
```java
public class MyException extends Exception{
  static final long serialVersionUID = -7034897193246939L;
  public MyException(){
    
  }
  public MyException(String msg){
    super(msg);
  }
}
```

#### 3. 自定义异常类的总结
直接继承RuntimeException即可，方法直接使用父类的。自定义异常类的意义在于异常类的名字，当异常出现时，可以根据名字判断异常类型。